London10-Afsha_Hossain_TV_dom_project

Afsha10 opened the pull request:
Level 300 WIP

Afsha10 committed:
Added full name and user name to the index.html file

Afsha10 committed:
Personalised index.html title

Afsha10 committed:
Personalised index.html title

Afsha10 committed:
Updated index.html and style.css and wrote pseudocode in script.js

Afsha10 committed:
Updated index.html, script.js and style.css by moving away from html and using the script.js to get the data from episode.js file

Restructured the index.html and changed the names a bit. We moved on from the using html to create the webpage. Instead converted the html elements to JS by creating variables  in JS and appended the variables inside other variables so that it represents the our html tree.
Also made changes to the pseudocode and css. Added the source in the footer.
To complete level 100, I need to remove the <p> from the episode descriptions and also made the episodeInfoCards fit in the cardsContainer. And probably several other things...

Afsha10 committed:
Removed the <p> from the beginning and </p> from the end of the episode description

Afsha10 committed:
Changed the body and cardsContainer background colour and added border to fix the layout

Afsha10 committed:
Trying to solve level 100 with only display flex

Afsha10 committed:
Completed level 100 using flex box

Afsha10 committed:
Update style.css

Afsha10 committed:
Update style.css

Changed .episode-name-num-holder width size to 20rem in css.

Afsha10 committed:
Update script.js

Correcting the last comment in script.js file

Afsha10 committed:
Almost completed level 100 using the grid method.

Need to find a way to stop the description text  going out of the episodeInfoHolder.

Afsha10 committed:
Updated index.html, style.css pseudocode script.js

Afsha10 committed:
Clearing unnecessary code

Afsha10 committed:
Almost done with level 100

Just need to make every the episode container size same.

Afsha10 committed:
Completed Level 100

Afsha10 committed:
Completed level 100

Afsha10 committed:
Deleted unnecessary files

Afsha10 committed:
Started working on level 200

Added some divs using JS particularly around images to have a cage for wild images. Added a bit of pseudocode for level 200. Corrected the previous h1 and h2 elements to span and p elements. Created input element for the search bar.

Afsha10 committed:
made some minor changes like changing div to span in JS and changes in CSS

Afsha10 committed:
Tried to make the search bar using scritp.js but failed

Afsha10 committed:
Cleared a few things

Afsha10 committed:
Tried to refactor to complete level 200

Afsha10 committed:
Corrected my HTML pseudocode

Afsha10 committed:
Completed level 200

Afsha10 committed:
Tried to do level 300. Updated index.html, script.js, style.css and psuedocode.md

Afsha10 committed:
Refactored script.js by creating a separate function for level 300 instead of doing all in one big function

Afsha10 committed:
Created Id for each episode in the drop down menu and tried to create an id for each of the cards to connect hem together using a "scrollIntoView" function

Afsha10 committed:
Almost completed level 300. Need to fix the overlapping top search bar covering the episode card tiles when an episode is  selecting in the drop down menu

maxbmaapc commented on script.js:
```diff
@@ -1,12 +1,160 @@
+  const episodeSelectHTML = document.querySelector("#select-html"); // this is the dropdown
+  const position = episodeSelectHTML.value; // we set a variable for the value which is the number
+  const episodeSelectHTMLId = "episode-card" + position; // position is from the dropdown list; we are making the id will will find
+  document.getElementById(episodeSelectHTMLId).scrollIntoView();
```
```suggestion
  document.getElementById(episodeSelectHTMLId).scrollIntoView({
    block: "center",
    behavior: "smooth",
  });
```
This should work. Header was overlapping because it's position is fixed. So now better

Afsha10 replied:
Your solution is so elegant!

Afsha10 committed:
Update script.js

Co-authored-by: Maksim Lukianenko <93668989+Mpanasetckiy@users.noreply.github.com>

Afsha10 committed:
Made some minor changes in html and css to make it a bit user friendly

Added an h1 element as a title of the page so the users understand what the page is about.

Added "All episodes" in the select menu bar

Added a placeholder to direct the users to search episode(s)

Afsha10 committed:
Updated script.js and used fetch method to get the API data

Updated script.js and used fetch method to get the API data, but the search is not working.

Afsha10 committed:
Renamed my variables for better understanding

Renamed the variables in script.js with HTML tags with "html" and changed a couple of id names in index.html

Afsha10 committed:
Create layout-with-var-names-id-class.jpg

Added a hand drawn layout of my page with variable names and id or class names

Afsha10 committed:
Update script.js

Tried to change the number of episodes displayed in searchResultDisplayHtml variable, but not working

maxbmaapc commented on script.js:
```diff
@@ -1,12 +1,156 @@
+// Global variables
+let allEpisodes;
+
+// Fetch episodes function
```
In your fetch function you simply forgot to assign your result to allEpisodes array, hence reference error when filtering :)

Afsha10 replied:
Good spot Maks!

maxbmaapc commented on script.js:
```diff
@@ -1,12 +1,156 @@
 function setup() {
-  const allEpisodes = getAllEpisodes();
-  makePageForEpisodes(allEpisodes);
+  fetch("https://api.tvmaze.com/shows/82/episodes")
+    .then(function (response) {
+      return response.json();
+    })
+    .then((result) => {
+      makePageForEpisodes(result);
+      buildEpisodeDropdownList(result);
+    });
 }
```
```suggestion
 const response = await fetch("https://api.tvmaze.com/shows/82/episodes");
  allEpisodes = await response.json();
  makePageForEpisodes(allEpisodes);
  buildEpisodeDropdownList(allEpisodes);
};
```

maxbmaapc commented on script.js:
```diff
@@ -1,12 +1,156 @@
+let allEpisodes;
+
+// Fetch episodes function
 function setup() {
```
```suggestion
const setup = async () => {
```
I recommend to change simple fetch for async/await

Afsha10 replied:
Thank you for telling me about async and await. I will try to learn about them and incorporate them in my code as you said !

maxbmaapc commented on script.js:
```diff
@@ -1,12 +1,156 @@
+  });
+
+  // Display searchedEpisodes
+  if (filteredEpisodes.length !== 0) {
```
Here is a small issue, do not remember whose Idea was to display filtered episodes only if we have them. I think better to display what we have, if we have none - display none. What u think?

Afsha10 replied:
That makes sense. We don't display if we don't have any episode. I will need to correct it.

maxbmaapc commented on script.js:
```diff
@@ -1,12 +1,156 @@
+  const position = episodeSelectHTML.value; // we set a variable for the value which is the number
+  // displaySearchedEpisodes(episodeSelectHTML);
+  const episodeSelectHTMLId = "episode-card" + position; // position is from the dropdown list; we are making the id will will find
+  document.getElementById(episodeSelectHTMLId).scrollIntoView({
```
Good scroll to view, but there are problems. When we display filtered episodes and select the episode, we get mistakes. Since there are several ways to show episodes, I recommend displaying selected episode via filter, this way we will avoid this issue

Afsha10 replied:
You are right. I will need to fix that too.

Afsha10 committed:
Made the search bar work

Made the search bar work by adding:  allEpisodes = result;
to the .then block inside the fetch function

Afsha10 committed:
Completed level 350 and correcting the "undefined"  part in the search display

Before I was using cardsContainerHtml.length but cardsContainerHtml is a DOM Element and it doesn’t have a length.  So I replaced cardsContainerHtml.length with index + 1

Afsha10 committed:
Corrected some bugs from 300

Afsha10 committed:
Update index.html

Added the autocomplete="off" attribute to the input element in index.html

Afsha10 committed:
Completed Level 350

Afsha10 committed:
Level 400 almost done with some bugs

Afsha10 committed:
Completed 400

Removed medium is null bug in line 89

Afsha10 committed:
added demo.html and demo.css

Afsha10 committed:
Update script.js

Changed "event" to "tvShowId"

Afsha10 committed:
Testing CSS Grid to get the new template

Trying level 500 with basic html and grid method

Afsha10 committed:
minor changes to demo files

Afsha10 committed:
Level 500 WIP

We created a makePageForShows  function and managed to get the showing appear at the start.

Afsha10 committed:
Creating a demo- WIP

Working on a demo just by using html and css grid

Afsha10 committed:
Added very basic info

Brought the name, rating, genre and runtime in the show info box without much details or style.

Afsha10 committed:
Renamed an element

Afsha10 committed:
Update script.js

Fixed the status info bug and added the labels inside the info box.

Afsha10 committed:
Update script.js

We have added span element for description and value for rating inside a p element rating container and appended it to the basic info container

Afsha10 committed:
Worked on the elements inside showBasicInfoContainerHtml

Worked on the elements inside showBasicInfoContainerHtml

Afsha10 committed:
Made some styling changes

Corrected an error in showStatusContainer and made some styling changes

Afsha10 committed:
Changed element names and html layout and css styling according to mobile first design

Tried to build mobile first layout for the tile layout using CSS

Afsha10 committed:
Made show displayed in the mobile version look better and cleaned up some unnecessary comments

Made some changes to css styling and cleaned up some unnecessary comments

Afsha10 committed:
Fixed the mobile version for displaying shows using css grid and flexbox

Afsha10 committed:
Made the pages with for show and episodes responsive

Created two different cards-container- show-cards-container and episode-cards-container. Used both flex box and css grid to get the desired outcome.

Afsha10 committed:
Fixed the displaySearchEpisode function

Fixed the displaySearchEpisode function by compensating for missing name and summary

Afsha10 committed:
Fixed the jumpToEpisode function

The problem was the selectEpisodeHtml.value was giving us a number which we could not compare with the meta data from the episodeList for eg. episode.name.
Solution: By changing the episodeOptionHtml.value = episodeList[i].name nested inside the episodeSelectedHtml have fixed the bug.

Afsha10 committed:
Minor updates to script.js and style.css

Changed function name and episode card background colour.

Afsha10 committed:
Made the button "View show listing" work

Added a navigation link to enable the user to return to the "shows listing" by the button "View Show Listing" is clicked.
When this is clicked and taken to the page with Show cards, the episodes listing is hidden.

maxbmaapc commented on style.css:
```diff
@@ -1,3 +1,361 @@
+  gap: 3rem;
+  justify-items: center;
+  justify-content: center;
+  width: 18.25rem;
```
```suggestion
  width: 18.25rem; // Use relative width like %
```

Afsha10 replied:
Thanks Maks!

maxbmaapc commented on style.css:
```diff
@@ -1,3 +1,361 @@
+  flex-direction: column;
+  border: 0.1rem #84d8ec solid;
+  background-color: #D1EBF1;
+  width: 18rem;
```
```suggestion
  width: 18rem; // Same here use % of the parent div
```

maxbmaapc commented on style.css:
```diff
@@ -1,3 +1,361 @@
+}
+
+
+@media screen and (min-width: 480px) {
```
This is mobile-first approach, is that what is intended here?

Afsha10 replied:
Yes, you are right. It is what I intended here.

maxbmaapc commented on style.css:
```diff
@@ -1,3 +1,361 @@
+    .show-card {
+    padding-bottom: 1rem;
+    gap: 0;
+    width: 80rem;
```
The problem in your rem measurements. Better to use %
```suggestion
    width: 100%;
```

Afsha10 committed:
Made the episodes page responsive

Made the episodes page responsive by using % for width instead of rem

Afsha10 committed:
Made the show page responsive

Made the show page responsive by using the flexbox instead of grid to arrange the elements inside the show cards. To use the flexbox, I created a new div container called "showImageSummaryInfoContainer" and I've put three other elements inside this container in a row.

Afsha10 committed:
Changed how showGenre innerHtml is written, bolded infoDescriptions in css and changed variable names

Made some the 4 small info description  bold using CSS, gave comma and spacing between different genres using .split and .join and changed some variable names.

Afsha10 committed:
Created the function displaySearchedShows and connected with the search bar

Created a function to be able to search shows in the search bar. It is same as episodes. I only changed the episodes to shows.

Afsha10 committed:
Fixed switching appearing of unnecessary dropdown menus while switching between shows and episodes

Fixed switching appearing of unnecessary dropdown menus while switching between shows and episodes by hiding visibility and allowing visibility.

Afsha10 committed:
Changed variable name created two event listeners and created an infinite loop

Afsha10 committed:
Corrected the infinite loop

Corrected the infinite loop by taking the eventlistener in line 211 outside the for loop

Afsha10 committed:
Created 2 separate search bars for episode and shows

Created 2 separate search bars- one for episodes and the other for shows. W make them visible and hide them same like we did with episode and show dropdowns in makePageForEpisodes and makePageForShows

Afsha10 committed:
Corrected some styling in style.css

Afsha10 committed:
Update style.css

Afsha10 committed:
Fixed a bug and think I completed level 500

Cleared the search input entered in show page and episode by episodeSearchInputHtml.value = "" where makePageForShow is called and by showSearchInputHtml.value = "" where makePageForEpisode is called.

Afsha10 committed:
Minor styling changes

Afsha10 committed:
Removed unnecessary text in Html for displaying search results as I have created it in JS.

faisalamin0206 commented on script.js:
```diff
@@ -1,12 +1,440 @@
 function setup() {
-  const allEpisodes = getAllEpisodes();
-  makePageForEpisodes(allEpisodes);
+  allShows = getAllShows().sort((a, b) => {
```
I really like all the variable naes you have made. Very clear and concise!

Afsha10 replied:
Thank you, Faisal!

faisalamin0206 commented on script.js:
```diff
@@ -1,12 +1,440 @@
+
+  episodeSearchResultDisplayHtml.style.display = "inline";
+  showSearchResultDisplayHtml.style.display = "none"; // hide episode search result from show page
+
```
It looks like you're doing the same thing 6 times which is very repetitive. Is there possibly a good word to possibly describe what we're doing. Is there a function name that we can use, using that word.

Afsha10 replied:
Thank you very much for pointing this out! Is the word to display and hide called called "toggle"?

faisalamin0206 replied:
Hi Afsha,
That is up to you. What I would say is that you should try it and see how it goes with "toggle()" and if you need any more help let myself or one of the team know :)

Afsha10 replied:
I might be able to manage this one. There are a lot of other bug I need to fix. It feels overwhelming sometimes. Thank you for you so much for your kind offer!

faisalamin0206 replied:
No problem! Trust me I am still learning as well so do not worry one bit. A bit of advice I can give in terms of feeling overwhelmed is to try and tell someone or write down al the problems you are having as you can tend to find the answer while you're speaking about the problem.

i hope this helps and let me know if you need anything else!

Afsha10 replied:
Hey Faisal, 
I have made some changes to reduce some repetition by using a toggleVisibility function. 
For me there is nothing like speaking about the problem. It gives me clarity. However, rarely I get to speak about the coding problems. I have been making checklists about the all the coding related problems in a notebook and have made some progress. Thank you for the reminder!
I am currently struggling a few small assignments related to fetch and API and we are starting React. Would you be interested in to pair programming sometime?

faisalamin0206 replied:
Hi Afsha,
I apologise for the late reply. I did not see this until today when I logged into Gituhub. Yes I would be happy to start pair programming. I am not super strong in programming as I have not done much since I finished my degree however, I still practice in my free time. I would be happy to do it. Let me know when you want to do it.

Afsha10 replied:
Hey Faisal!
No worries about the delay. Before we start pair programming, it would be great if you could join our in-person training sessions on Saturdays. It'll give us a chance to meet face-to-face and discuss pair programming and our projects in more detail.
Let me know if you're interested, and I'll share the details of the upcoming Saturday session. You can also find the information in our Slack channel. Feel free to check it out.

faisalamin0206 replied:
Hi Afsha,
Yes I will be more than happy to come in person for a session. I won't be able to attend tomorrow as I am away for the weekend however nest Saturday works perfectly for me. Could you please forward me the details and I will see you then!

Afsha10 replied:
Tomorrow it will be remote. Next week it will probably be at Lego, which is exciting! Feel free to DM on slack or here if you need the address.

illicitonion commented on script.js:
```diff
@@ -1,12 +1,440 @@
+  episodeList.forEach((episode, index) => {
+    // Get the episode name, season number, episode number, and combine them
+    const episodeName = episode.name;
+    const paddedSeasonNum = ("0" + episode.season).slice(-2);
```
Rather than repeating this code on this and the next line, maybe we could extract a function with a clear name which shows what it's doing?

Afsha10 replied:
Thank you for pointing this out! I created 2 different functions to format the name in the episode dropdown list and to format episode titles in episode cards. I had to create two functions because they are formatted slightly differently.

illicitonion commented on script.js:
```diff
@@ -1,12 +1,440 @@
+  // Display selectedEpisode
+  rootHtml.innerHTML = "";
+  makePageForEpisodes(selectedEpisode);
+  // buildEpisodeDropdownList(selectedShows);
```
I'd recommend deleting the commented  code, it makes it easier to follow what's actually going on (and easier to review!)

Afsha10 replied:
Apologies for that. I have removed those comments and the unnecessary console logs.

illicitonion commented on script.js:
```diff
@@ -1,12 +1,440 @@
+
+    // creating a status element and its value inside showBasicInfoContainerHtml
+
+    const showStatusValueHtml = document.createElement("span");
```
I'm seeing a pattern of a lot of "document.createElement ; element.classList.add ; element.innerHTML =" (and also then `appendChild`) - is there a useful function to extract here which does these three things? It may make it more clear to see what's going on.

Afsha10 replied:
You are right about the pattern. Will it be something like createElementAddClassAppend function? 
I'm not sure about the innerHTML. Sometimes we use .innerHTML and sometimes we use .src.

illicitonion reviewed (commented):
This is generally looking really solid, nicely put together!

One usability thing I noticed using your page - if you filter by searching in the text box, the page moves around a bit - can you work out why this is? Can you think how to improve this?

Afsha10 committed:
Created a function called toggleVisibility to change visibility of show and episode elements

Used a function to show and hide episode and show elements when the show and episode pages are displayed

Afsha10 commented:
> This is generally looking really solid, nicely put together!
> 
> One usability thing I noticed using your page - if you filter by searching in the text box, the page moves around a bit - can you work out why this is? Can you think how to improve this?

Afsha10 committed:
Refactored by creating 2 different functions to format the episodeTitles in the episodeCards and episodeTitles  in episodeDropdownMenu

Created 2 different functions to format the name in the episode dropdown list and also episode titles in episode cardsRemoved unnecessary comments and console.logs.

Afsha10 committed:
Fixed the episode and show pages to appear at the top the of page

Fixed the episode and show pages to appear at the top the of page by using  this code: window.scrollTo(0, 0) ;

illicitonion commented on script.js:
```diff
@@ -1,12 +1,442 @@
+}
+
+function getFormattedEpisodeNameInEpisodeCards(episode) {
+  const paddedSeasonNum = ("0" + episode.season).slice(-2);
+  const paddedEpisodeNum = ("0" + episode.number).slice(-2);
```
This looks a lot like you're implementing the same thing in this function and in `getFormattedEpisodeNameInDropdown` - is there a third function you could extract, which you could call from both of those, to avoid having to write this twice?

illicitonion commented on script.js:
```diff
@@ -1,12 +1,442 @@
-  const rootElem = document.getElementById("root");
-  rootElem.textContent = `Got ${episodeList.length} episode(s)`;
+  // make the episode dropdown list visible and hide the show dropdown list
+  toggleVisibility(episodeSelectHtml, showSelectHtml);
```
"toggle" implies "swap" (i.e. that if you call toggle once it will change something, and call it again it will change it back) - I think what you're really doing in `toggleVisibility` is _hiding_ the elements - can you think of a more clear name for the function?

illicitonion commented on script.js:
```diff
@@ -1,12 +1,442 @@
+
+    const episodeNameNumSeasonComboHolderHtml = document.createElement("div");
+    const episodeNameNumSeasonComboTextHtml = document.createElement("span");
+    episodeNameNumSeasonComboHolderHtml.classList.add(
```
It looks like every element you create, you add a class-name for - can you maybe extract a function which would mean you could write something like:
```js
const episodeNameNumSeasonComboHolderHtml = createElement("div", "episode-name-num-holder");
```

instead of keep having to the same things over and over?

Are there more things you're repeatedly doing you could _also_ extract into this function, as well?

illicitonion reviewed (commented):
This is looking really well put together - good job!
